### PR TITLE
[Docker] Update the Dockerfile to use up-to-date instructions for Ubuntu

### PR DESCRIPTION
**Pull request by rexwangcc**
Related issue = #

(I recently wanted to use taichi through a ssh tunnel on a container cluster and realized the Dockerfile is outdated)

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

## Review Instructions:

In order to test this out, please either build the image locally with: `docker build -t taichi:2b181a26 .`(this builds the image from source so may take up to 40 mins) OR pull the image I built and pushed with : `docker run -it rexwangcc/taichi:2b181a26`

- For a quick check, you could run `ti test` within the container. 

- For a more sophisticated test, with `docker`installed properly, please follow:
### On macOS (cpu only)
1. Make sure XQuartz and socat are installed:
```
brew cask install xquartz
brew install socat
```
2. Temporally disable the xhost access-control: `xhost +`
3. Run an example, such as mpm88 with: `docker run -it -e DISPLAY=$(ipconfig getifaddr en0):0 rexwangcc/taichi:2b181a26 ti example mpm88`
4. Re-enable the xhost  access-control: `xhost -`

### On Ubuntu (cuda support)
1. Install NVIDIA container runtime for docker:
```
curl -s -L https://nvidia.github.io/nvidia-container-runtime/gpgkey | sudo apt-key add -
distribution=$(. /etc/os-release;echo $ID$VERSION_ID)
curl -s -L https://nvidia.github.io/nvidia-container-runtime/$distribution/nvidia-container-runtime.list |\
    sudo tee /etc/apt/sources.list.d/nvidia-container-runtime.list
sudo apt-get update
sudo apt-get install nvidia-container-runtime
```
2. Restart docker daemon:
```
sudo systemctl stop docker
sudo systemctl start docker
```
3. Make sure xorg is installed: `sudo apt-get install xorg`
4.  Temporally disable the xhost access-control: `xhost +`
5. Run an example, such as mpm88 with: `sudo docker run -it --gpus all -e DISPLAY=$DISPLAY -v /tmp/.X11-unix:/tmp/.X11-unix rexwangcc/taichi:2b181a26 ti example mpm88`
6. Re-enable the xhost  access-control: `xhost -`

## Open Questions:

- The Dockerfile this PR updates is effectively a `Dockerfile.dev` which fullfills my own requirements (remote taichi development), But would it bring any value to end-users to have a runtime `Dockerfile` separately?
- Would it be possible to have a repository on Docker Hub so we could set up a free build trigger on Docker Hub side to automatically build tagged images when a new release is cut on Github?

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,73 +1,60 @@
...
+RUN make install
+
+# Install Taichi from source
+WORKDIR /taichi-dev
+RUN git clone https://github.com/taichi-dev/taichi --depth=1 --branch=master
+RUN cd taichi && \
+    git submodule update --init --recursive --depth=1 && \
+    mkdir build
+WORKDIR /taichi-dev/taichi/build
+RUN cmake .. -DPYTHON_EXECUTABLE=python3 -DCUDA_VERSION=10.0 -DTI_WITH_CUDA:BOOL=True
```
Comment (yuanming-hu) [inline Dockerfile:50]:
```suggestion
RUN cmake .. -DPYTHON_EXECUTABLE=python3 -DTI_WITH_CUDA:BOOL=True
```

We no longer need to specify `CUDA_VERSION` - sorry about the confusing documentation. It should be updated after #1388

**Comment by rexwangcc** [comment]
> Thank you! This is very helpful. I can't wait to try it on my own - do I need something like https://github.com/NVIDIA/nvidia-docker? It would be great to have brief documentation on how to use the `Dockerfile` :-)

Thanks for the review! In order to make the CUDA work on Ubuntu machines, yes you would need to install the nvidia-docker kit (see the review instructions). And sorry for the missing docs on this PR, I'm planning to copy-paste the review instructions to the dev-install doc, I haven't tried this with Docker on Windows so only macOS and Ubuntu instructions will be available for now, if you don't mind.

**Comment by yuanming-hu** [comment]
Thank you so much! Works like a charm!

Screenshot from 2020-07-04 17-04-31

(Sorry that I forgot to read your PR description carefully.)



> I'm planning to copy-paste the review instructions to the dev-install doc, I haven't tried this with Docker on Windows so only macOS and Ubuntu instructions will be available for now, if you don't mind.

That sounds great! Not having Windows is totally fine for now.

**Comment by yuanming-hu** [comment]
I added `Docker` to PR tags. Feel free to merge if that looks good to you :-)

**Comment by rexwangcc** [comment]
> I added `Docker` to PR tags. Feel free to merge if that looks good to you :-)

Thanks! I added a Docker section to the dev documentation. Will merge the PR once the builds pass!

**Pull request merged**